model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -125,24 +81,250 @@ impl Serialize for CommandDataOption {
+                        let val = value_opt.ok_or_else(|| DeError::missing_field("value"))?;
+
+                        match val {
+                            ValueEnvelope::Integer(i) => {
+                                let id = RoleId::new(i.try_into().map_err(|_| {
+                                    DeError::custom("could not convert signed integer to unsigned")
+                                })?)
+                                .ok_or_else(|| DeError::custom("role id was zero"))?;
+
+                                CommandOptionValue::Role(id)
+                            }
+                            ValueEnvelope::Id(id) => CommandOptionValue::Role(RoleId(id.0)),
+                            other => {
+                                return Err(DeError::custom(&format!(
+                                    "expected integer got {:?}",
+                                    other
+                                )));
+                            }
+                        }
+                    }
+                    CommandOptionType::String => {
+                        let val = value_opt.ok_or_else(|| DeError::missing_field("value"))?;
+
+                        if let ValueEnvelope::String(s) = val {
+                            CommandOptionValue::String(s)
+                        } else {
+                            return Err(DeError::custom(&format!("expected string got {:?}", val)));
+                        }
+                    }
+                    CommandOptionType::SubCommand => CommandOptionValue::SubCommand(options),
+                    CommandOptionType::SubCommandGroup => {
+                        CommandOptionValue::SubCommandGroup(options)
+                    }
+                    CommandOptionType::User => {
+                        let val = value_opt.ok_or_else(|| DeError::missing_field("value"))?;
+
+                        match val {
+                            ValueEnvelope::Integer(i) => {
+                                let id = UserId::new(i.try_into().map_err(|_| {
+                                    DeError::custom("could not convert signed integer to unsigned")
+                                })?)
+                                .ok_or_else(|| DeError::custom("user id was zero"))?;
+
+                                CommandOptionValue::User(id)
+                            }
+                            ValueEnvelope::Id(id) => CommandOptionValue::User(UserId(id.0)),
+                            other => {
+                                return Err(DeError::custom(&format!(
+                                    "expected integer got {:?}",
+                                    other
+                                )));
+                            }
+                        }
+                    }
+                };
+
+                Ok(CommandDataOption { name, value })
+            }
+        }
+
+        deserializer.deserialize_any(CommandDataOptionVisitor)
```

Comment:
We know that this is a map, this follows other code style
```suggestion
        deserializer.deserialize_map(CommandDataOptionVisitor)
```